File: game-engine/src/main/java/org/luxons/sevenwonders/game/moves/Move.java
package org.luxons.sevenwonders.game.moves;

import java.util.List;

import org.luxons.sevenwonders.game.Settings;
import org.luxons.sevenwonders.game.api.PlayerMove;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.cards.Card;
import org.luxons.sevenwonders.game.resources.ResourceTransactions;

public abstract class Move {

    private int playerIndex;

    private Card card;

    private MoveType type;

    private ResourceTransactions transactions;

    Move(int playerIndex, Card card, PlayerMove move) {
        this.playerIndex = playerIndex;
        this.card = card;
        this.type = move.getType();
        this.transactions = new ResourceTransactions(move.getTransactions());
    }

    public int getPlayerIndex() {
        return playerIndex;
    }

    public Card getCard() {
        return card;
    }

    public MoveType getType() {
        return type;
    }

    public ResourceTransactions getTransactions() {
        return transactions;
    }

    public abstract void validate(Table table, List<Card> playerHand) throws InvalidMoveException;

    public abstract void place(Table table, List<Card> discardedCards, Settings settings);

    public abstract void activate(Table table, List<Card> discardedCards, Settings settings);
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/moves/Move.kt
package org.luxons.sevenwonders.game.moves

import org.luxons.sevenwonders.game.PlayerContext
import org.luxons.sevenwonders.game.Settings
import org.luxons.sevenwonders.game.api.PlayerMove
import org.luxons.sevenwonders.game.cards.Card
import org.luxons.sevenwonders.game.resources.ResourceTransactions

internal abstract class Move(
    val move: PlayerMove,
    val card: Card,
    val playerContext: PlayerContext
) {
    val type: MoveType = move.type

    val transactions: ResourceTransactions = move.transactions

    abstract fun place(discardedCards: MutableList<Card>, settings: Settings)

    abstract fun activate(discardedCards: List<Card>, settings: Settings)
}

class InvalidMoveException internal constructor(move: Move, message: String) : IllegalArgumentException(
    "Player ${move.playerContext.index} cannot perform move ${move.type}: $message"
)

internal fun MoveType.resolve(move: PlayerMove, card: Card, context: PlayerContext): Move = when (this) {
    MoveType.PLAY -> PlayCardMove(move, card, context)
    MoveType.PLAY_FREE -> PlayFreeCardMove(move, card, context)
    MoveType.UPGRADE_WONDER -> BuildWonderMove(move, card, context)
    MoveType.DISCARD -> DiscardMove(move, card, context)
    MoveType.COPY_GUILD -> CopyGuildMove(move, card, context)
}
